app/routes/index.js

```diff
@@ -19,115 +19,82 @@ mongoose.connect('mongodb://localhost/todw', function(err) {
-	console.log(url);
-
-	var options = {
-	  url: url,
-	  headers: {
-	    'Accept': 'application/json'
-	  }
-	};
-
-	request(options, function (err, resp, body) {
-			// console.log("response headers: ", resp.headers);
-		if (!err && resp.statusCode == 200) {
-			res.set(resp.headers);
-			res.send(body);
-		} else {
-			res.send(err);
-		}
-	});
+    // console.log('The sessionKey = ', accessToken.sessionKey());
+    var url = coursesEndpoint + req.path +
+                '?per_page=100&access_token=' + accessToken.sessionKey() + appendQueryParams(req.originalUrl);
+    console.log(url);
+
+    var options = {
+      url: url,
+      headers: {
+        'Accept': 'application/json'
+      }
+    };
+
+    request(options, function (err, resp, body) {
+            // console.log("response headers: ", resp.headers);
+        if (!err && resp.statusCode == 200) {
+            res.set(resp.headers);
+            res.send(body);
+        } else {
+            res.send(err);
+        }
+    });
 });
 
 router.get('/api/csp-framework', function (req, res) {
     res.send(cspFramework);
 });
-	
 
-		// api ---------------------------------------------------------------------
-	// get all the unit and the lessons
-router.get('/api/units', function(req, res) {
+function respondWithQuery (fn, successCode) {
+    successCode || (successCode = 200);
 
-		// use mongoose to get unit with the canvasUnitID in the database
-		console.log('/api/units');
-	UnitMetadata.find(function(err, unitMetadata) {
+    return function (req, res, next) {
+        function onSuccess (result) {
+            res.status(successCode);
+            if (result && successCode !== 204) {
+                res.json(result);
+            } else {
+                res.end();
+            }
+        }
 
-				// if there is an error retrieving, send the error. nothing after res.send(err) will execute
-		if (err)
-			res.send(err)
-
-		console.log(unitMetadata);
-		res.json(unitMetadata); // return all units in JSON format
-	
-	});
-
-});
-
-	// create todo and send back all todos after creation
-router.post('/api/unit', function(req, res) {
-
-	// create a todo, information comes from AJAX request from Angular
-	LessonItem.create({
-		text : req.body.text,
-		done : false
-	}, function(err, lessonItem) {
-		if (err)
-			res.send(err);
-
-		// get and return all the todos after you create another
-		getLessonItems(res);
-	});
-
-});
-
-/* PUT /units/:id */
-router.put('/:id', function(req, res, next) {
-  Todo.findByIdAndUpdate(req.params.id, req.body, function (err, post) {
-    if (err) return next(err);
-    res.json(post);
-  });
-});
-
-	// delete a todo
-router.delete('/api/unit/:unit_id', function(req, res) {
-	Todo.remove({
-		_id : req.params.lessonItem_id
-	}, function(err, todo) {
-		if (err)
-			res.send(err);
+        fn(req).then(onSuccess, next);
+    }
+}
 
-		getTodos(res);
-	});
-});
+router.get('/api/units', respondWithQuery(function (req) {
```

Comment:
This should look a little different to you, since you're used to seeing
```javascript
router.get('/api/units', function (req, res, next) {
  someQuery(function (err, data) {
    if (err) {
      next(err);
    } else {
      res.json(data);
    }
  });
});
```
The `respondWithQuery()` function makes it easy to repeat this pattern.  Returning a function from a function (as it does) is a bit strange at first, but at least for now you can see from the code below how much easier it is to define the routes.  More specifically, I don't have to repeat the same error-handling and response-generating code for each request.